core/srp.js

```diff
@@ -1,226 +1,678 @@
+                                         this._publicA.toBits(), this._publicB.toBits());
+    if (u.mod(group.N).equals(0)) {
+      throw new sjcl.exception.corrupt("u mod N == 0! SRP must be aborted!");
+    }
+    this._u = u;
+
+    /* S = (B - kg^x) ^ (a + ux) */
+    k = group.calculateK(this._hash);
+    x = this._calculateX(this._salt);
+    v = this._calculateV(this._salt, x);
+    S = this._publicB.sub(v.mulmod(k, group.N));
+    S = S.powermod(this._secretA.add(u.mulmod(x, group.N)), group.N);
+    this._S = S;
+
+    this._K = this._hash.hash(this._S.toBits());
+    return this._K;
+  },
+
+  /**
+  * Calculate client authentication message.
+  * @return {bitArray} SRP M1 client authentication
+  */
+  getClientAuth: function() {
+    this._M1 = sjcl.keyex.srp._getAuth1(this._hash, this._group, this.username, this._salt,
+                                              this._publicA.toBits(), this._publicB.toBits(), this._K);
+    return this._M1;
+  },
+
+  /**
+  * Authenticate server.
+  * @param {bitArray} serverM Server authentication message (optional).
+  * @return {Boolean} True indicates authentication success and False failure.
+  */
+  authenticateServer: function(serverM) {
+    if (typeof serverM !== "object" || sjcl.bitArray.bitLength(serverM) == 0) {
+      throw new sjcl.exception.invalid("serverM must be a bitArray!");
+    }
+    this._M2 = sjcl.keyex.srp._getAuth2(this._hash, this._publicA.toBits(), this._M1, this._K);
+    this.authenticated = sjcl.bitArray.equal(serverM, this._M2);
+    return this.authenticated;
+  },
+
+  /**
+  * Calculate SRP x
+  * @param {bitArray} salt User salt.
+  * @return {sjcl.bn} SRP x.
+  */
+  _calculateX: function(salt) {
+    var password_hash, x;
+
+    password_hash = new this._hash();
+    password_hash.update(this.username);
+    password_hash.update(":");
+    password_hash.update(this._password);
+    password_hash = password_hash.finalize();
+
+    x = new this._hash();
+    x.update(salt);
+    x.update(password_hash);
+    return sjcl.bn.fromBits(x.finalize());
+  },
+
+  /**
+  * Calculate SRP v
+  * @param {bitArray} salt User salt.
+  * @return {sjcl.bn} SRP v.
+  */
+  _calculateV: function(salt, x) {
+    var group = this._group;
+
+    x = x || this._calculateX(salt);
     return group.g.powermod(x, group.N);
   },
 
   /**
-   * Calculates SRP x.
-   *   x = SHA1(<salt> | SHA(<username> | ":" | <raw password>)) [RFC 2945]
-   * @param {String} I The username.
-   * @param {String} P The password.
-   * @param {Object} s A bitArray of the salt.
-   * @return {Object} A bitArray of SRP x.
-   */
-  makeX: function(I, P, s) {
-    var inner = sjcl.hash.sha1.hash(I + ':' + P);
-    return sjcl.hash.sha1.hash(sjcl.bitArray.concat(s, inner));
+  * Test helper. Compare computed x against provided
+  * @param {bitArray} salt User salt.
+  * @param {bitArray} x SRP x
+  * @return {Boolean} True is equal.
+  */
+  testX: function(salt, x) {
```

Comment:
Options are: we expose private variables and functions which are needed for testing or we don't check X, U and S in tests.